src/extension.ts

```diff
@@ -4,148 +4,195 @@
-					reader: socket,
-					writer: socket
-				});
-
-				socket.on('end', () => console.log("Disconnected"));
-			}).on('error', (err) => {
-				// handle errors here
-				throw err;
-			});
-
-			// grab a random port.
-			server.listen(() => {
-				// Start the child java process
-				let options = { cwd: context.extensionPath };
-				let launcherPath = context.extensionPath + "/coursier"
-
-				let port = (server.address() as net.AddressInfo).port
-
-				let version = vscode.workspace.getConfiguration('smithyLsp').get("version", "`")
-
-				// Downloading latest poms
-				let resolveArgs = ["resolve", "--mode", "force", "com.disneystreaming.smithy:smithy-language-server:" + version, "-r", "m2local"]
-				let resolveProcess = child_process.spawn(launcherPath, resolveArgs, options)
-				resolveProcess.on('exit', exitCode => {
-					console.log("Exit code : " + exitCode)
-					if (exitCode == 0) {
-						console.log("Launching smithy-language-server version:" + version)
-
-						let launchargs = ["launch", "com.disneystreaming.smithy:smithy-language-server:" + version, "-r", "m2local", "--", port.toString()]
-
-						let childProcess = child_process.spawn(launcherPath, launchargs, options);
-
-						childProcess.stdout.on('data', (data) => {
-							console.log(`stdout: ${data}`);
-						});
-
-						childProcess.stderr.on('data', (data) => {
-							console.error(`stderr: ${data}`);
-						});
-
-						childProcess.on('close', (code) => {
-							console.log(`LSP exited with code ${code}`);
-						});
-					} else {
-						console.log(`Could not resolve smithy-language-server implementation`)
-					}
-				})
-
-				// Send raw output to a file
-				if (!fs.existsSync(context.storagePath))
-					fs.mkdirSync(context.storagePath);
-			});
-		});
-	};
-
-	// Options to control the language client
-	let clientOptions: LanguageClientOptions = {
-		// Register the server for plain text documents
-		documentSelector: [{ scheme: 'file', language: 'smithy' }, { scheme: 'smithyjar', language: 'smithy' }],
-		synchronize: {
-			// Notify the server about file changes to 'smithy-build.json' or 'smithy.json' files contained in the workspace
-			fileEvents: workspace.createFileSystemWatcher('**/{smithy-build, smithy}.json')
-		}
-	};
-
-	// Create the language client and start the client.
-
-	client = new LanguageClient(
-		'smithyLsp',
-		'Smithy LSP',
-		createServer,
-		clientOptions
-	);
-	const smithyContentProvider = createSmithyContentProvider(client);
-	context.subscriptions.push(workspace.registerTextDocumentContentProvider('smithyjar', smithyContentProvider));
-
-	// Start the client. This will also launch the server
-	client.start();
+    function createServer(): Promise<StreamInfo> {
+        function startServer(executable: string): Promise<StreamInfo> {
+            console.log(`Executable located at" ${executable}.`)
+            return new Promise((resolve, reject) => {
+                var server = net.createServer((socket) => {

```

Comment:
**suggestion (non-blocking):** can we use const here? worst case, let

Reply:
done